File: ios/stereoCamera/stereoCamera/Comm/Commands/SendStatus.swift
//
//  SendStatus.swift
//  stereoCamera
//
//  Created by hallmarklabs on 6/25/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class SendStatus : Command
{
    var status:Status

    init(_ status: Status)
    {
        self.status = status
        super.init()
        doInit()
    }
    
    override init()
    {
        self.status = Status.NONE
        super.init()
        doInit()
    }
    
    private func doInit()
    {
        cmdtype = CommandTypes.RECEIVE_STATUS
        expectsResponse = false
    }
    
    override func send(comm: Comm) -> Bool
    {
        let result = super.send(comm: comm)
        if (!result) { return false }
        
        var bytes:[UInt8] = []
        bytes += Bytes.toByteArray(status)
        
        let sz = comm.write(buf: bytes)
        return (sz <= 0) ? false : true
    }
    
    override func receive(comm: Comm) -> Bool
    {
        let result = super.receive(comm: comm)
        if (!result) { return false }
        
        let (buf, sz) = comm.read(sz: 1)
        if (sz <= 0) { return false }
        status = Bytes.fromByteArray(buf)
        
        return true
    }
}

File: android/app/src/main/java/com/munger/stereocamera/ip/command/commands/SendStatus.java
package com.munger.stereocamera.ip.command.commands;

import com.munger.stereocamera.fragment.PreviewFragment;
import com.munger.stereocamera.ip.command.Comm;
import com.munger.stereocamera.ip.command.Command;

import java.io.IOException;

public class SendStatus extends Command
{
	public PreviewFragment.Status status;

	public SendStatus(PreviewFragment.Status status)
	{
		super();
		this.status = status;
		doInit();
	}

	public SendStatus()
	{
		super();
		this.status = PreviewFragment.Status.NONE;
		doInit();
	}

	private void doInit()
	{
		cmdtype = Type.RECEIVE_STATUS;
		expectsResponse = false;
	}

	@Override
	public boolean send(Comm comm)
	{
		boolean result = super.send(comm);
		if (result == false)
			{ return false; }

		try
		{
			comm.putByte((byte) status.ordinal());
		}
		catch(IOException e){
			return false;
		}

		return true;
	}

	@Override
	public boolean receive(Comm comm)
	{
		boolean result = super.receive(comm);
		if (result == false)
			{ return false; }

		try
		{
			byte b = comm.getByte();
			PreviewFragment.Status[] statList = PreviewFragment.Status.values();

			if (b < 0 || b >= statList.length)
				return false;

			status = statList[b];

		}
		catch(IOException e){
			return false;
		}

		return true;
	}
}
